### ISSUE TITLE
schema attr editor

### ISSUE THREAD

**Issue by dustingetz** [hyperfiddle/hyperfiddle#330]

**Comment by dustingetz** [comment]
The readonly logic is broken, but the test proves the algebra is right. I think there is a downstream issue with readonly which could be related to the link readonly state being broken.

**Comment by dustingetz** [comment]
The checkbox dom actually is getting :read-only true, but it needs to be :disabled true, somewhere i lost the readonly->disabled adapter

**Comment by dustingetz** [comment]
https://github.com/hyperfiddle/hyperfiddle/commit/1a69429ad293494a4c43820b412721e57af06648

**Comment by dustingetz** [comment]
Its not finished though - attr editor is still sending through datoms that should be buffered while the attr is uncompleted

**Comment by dustingetz** [comment]
I think the react-context is giving us an old value from the context and not the new value of the context that was rebound.

**Comment by dustingetz** [comment]
No, confirmed working fine

**Comment by dustingetz** [comment]
This appears to be a clojurescript compiler bug with letfn

Two independent letfn bindings which bind the same symbol, are colliding. The second one gets called by both. Here is the compiler dump where we can see the rebinding.

```
// Compiled by ClojureScript 1.9.946 {}
goog.provide('hyperfiddle.ide.fiddles.schema_attribute');
goog.require('cljs.core');
goog.require('clojure.set');
goog.require('contrib.datomic_tx');
goog.require('contrib.reactive');
goog.require('contrib.reagent');
goog.require('hypercrud.browser.context');
goog.require('hyperfiddle.ui');
hyperfiddle.ide.fiddles.schema_attribute.special_attrs = new cljs.core.PersistentHashSet(null, new cljs.core.PersistentArrayMap(null, 3, [new cljs.core.Keyword("db","valueType","db/valueType",1827971944),null,new cljs.core.Keyword("db","cardinality","db/cardinality",-104975659),null,new cljs.core.Keyword("db","ident","db/ident",-737096),null], null), null);
hyperfiddle.ide.fiddles.schema_attribute.completed_QMARK_ = (function hyperfiddle$ide$fiddles$schema_attribute$completed_QMARK_(entity){
return clojure.set.subset_QMARK_.call(null,hyperfiddle.ide.fiddles.schema_attribute.special_attrs,cljs.core.set.call(null,cljs.core.keys.call(null,entity)));
});
hyperfiddle.ide.fiddles.schema_attribute.read_only_QMARK_ = (function hyperfiddle$ide$fiddles$schema_attribute$read_only_QMARK_(k,record){
if(cljs.core.truth_(hyperfiddle.ide.fiddles.schema_attribute.special_attrs.call(null,k))){
return false;
} else {
return cljs.core.not.call(null,hyperfiddle.ide.fiddles.schema_attribute.completed_QMARK_.call(null,record));

}
});
hyperfiddle.ide.fiddles.schema_attribute.merge_in_tx = (function hyperfiddle$ide$fiddles$schema_attribute$merge_in_tx(entity,tx,ctx){
return cljs.core.reduce.call(null,(function (entity__$1,p__54819){
var vec__54820 = p__54819;
var op = cljs.core.nth.call(null,vec__54820,(0),null);
var e = cljs.core.nth.call(null,vec__54820,(1),null);
var a = cljs.core.nth.call(null,vec__54820,(2),null);
var v = cljs.core.nth.call(null,vec__54820,(3),null);
var valueType = cljs.core.deref.call(null,contrib.reactive.cursor.call(null,new cljs.core.Keyword("hypercrud.browser","schemas","hypercrud.browser/schemas",-1320254913).cljs$core$IFn$_invoke$arity$1(ctx),new cljs.core.PersistentVector(null, 4, 5, cljs.core.PersistentVector.EMPTY_NODE, ["$",a,new cljs.core.Keyword("db","valueType","db/valueType",1827971944),new cljs.core.Keyword("db","ident","db/ident",-737096)], null)));
var v__$1 = ((cljs.core._EQ_.call(null,new cljs.core.Keyword("db.type","ref","db.type/ref",-1728373079),valueType))?new cljs.core.PersistentArrayMap(null, 1, [new cljs.core.Keyword("db","id","db/id",-1388397098),v], null):v);
var G__54823 = op;
var G__54823__$1 = (((G__54823 instanceof cljs.core.Keyword))?G__54823.fqn:null);
switch (G__54823__$1) {
case "db/add":
return cljs.core.assoc.call(null,entity__$1,a,v__$1);

break;
case "db/retract":
return cljs.core.dissoc.call(null,entity__$1,a);

break;
default:
throw (new Error(["No matching clause: ",cljs.core.str.cljs$core$IFn$_invoke$arity$1(G__54823__$1)].join('')));

}
}),cljs.core.into.call(null,cljs.core.PersistentArrayMap.EMPTY,entity),tx);
});
var user_with_BANG_ = (function hyperfiddle$ide$fiddles$schema_attribute$user_with_BANG_(f_BANG_,special_attrs_state,ctx,tx){
cljs.core.println.call(null,"38");

var entity = cljs.core.deref.call(null,hypercrud.browser.context.entity.call(null,ctx));
var new_entity = hyperfiddle.ide.fiddles.schema_attribute.merge_in_tx.call(null,entity,tx,ctx);
var G__54826 = new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [hyperfiddle.ide.fiddles.schema_attribute.completed_QMARK_.call(null,entity),hyperfiddle.ide.fiddles.schema_attribute.completed_QMARK_.call(null,new_entity)], null);
if(cljs.core._EQ_.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [false,false], null),G__54826)){
return cljs.core.swap_BANG_.call(null,special_attrs_state,contrib.datomic_tx.into_tx,tx);
} else {
if(cljs.core._EQ_.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [false,true], null),G__54826)){
f_BANG_.call(null,contrib.datomic_tx.into_tx.call(null,cljs.core.deref.call(null,special_attrs_state),tx));

return cljs.core.reset_BANG_.call(null,special_attrs_state,null);
} else {
if(cljs.core._EQ_.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [true,false], null),G__54826)){
return f_BANG_.call(null,tx);
} else {
if(cljs.core._EQ_.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [true,true], null),G__54826)){
return f_BANG_.call(null,tx);
} else {
throw (new Error(["No matching clause: ",cljs.core.str.cljs$core$IFn$_invoke$arity$1(G__54826)].join('')));

}
}
}
}
});
hyperfiddle.ide.fiddles.schema_attribute.build_valueType_and_cardinality_renderer = (function hyperfiddle$ide$fiddles$schema_attribute$build_valueType_and_cardinality_renderer(special_attrs_state){
return contrib.reagent.from_react_context.call(null,(function (p__54827,value){
var map__54828 = p__54827;
var map__54828__$1 = ((((!((map__54828 == null)))?((((map__54828.cljs$lang$protocol_mask$partition0$ & (64))) || ((cljs.core.PROTOCOL_SENTINEL === map__54828.cljs$core$ISeq$)))?true:false):false))?cljs.core.apply.call(null,cljs.core.hash_map,map__54828):map__54828);
var ctx = cljs.core.get.call(null,map__54828__$1,new cljs.core.Keyword(null,"ctx","ctx",-493610118));
var props = cljs.core.get.call(null,map__54828__$1,new cljs.core.Keyword(null,"props","props",453281727));
return new cljs.core.PersistentVector(null, 5, 5, cljs.core.PersistentVector.EMPTY_NODE, [contrib.reagent.fix_arity_1_with_context,hyperfiddle.ui.hyper_control.call(null,ctx),value,cljs.core.assoc.call(null,ctx,new cljs.core.Keyword(null,"user-with!","user-with!",2055209278),contrib.reactive.partial.call(null,user_with_BANG_,contrib.reactive.partial.call(null,cljs.core.println,"63 "),special_attrs_state,ctx)),props], null);
}));
});
var user_with_BANG_ = (function hyperfiddle$ide$fiddles$schema_attribute$user_with_BANG_(f_BANG_,special_attrs_state,ctx,tx){
cljs.core.println.call(null,"67");

var entity = cljs.core.deref.call(null,hypercrud.browser.context.entity.call(null,ctx));
var new_entity = hyperfiddle.ide.fiddles.schema_attribute.merge_in_tx.call(null,entity,tx,ctx);
var G__54831 = new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [hyperfiddle.ide.fiddles.schema_attribute.completed_QMARK_.call(null,entity),hyperfiddle.ide.fiddles.schema_attribute.completed_QMARK_.call(null,new_entity)], null);
if(cljs.core._EQ_.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [false,false], null),G__54831)){
return f_BANG_.call(null,tx);
} else {
if(cljs.core._EQ_.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [false,true], null),G__54831)){
f_BANG_.call(null,contrib.datomic_tx.into_tx.call(null,cljs.core.deref.call(null,special_attrs_state),tx));

return cljs.core.reset_BANG_.call(null,special_attrs_state,null);
} else {
if(cljs.core._EQ_.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [true,false], null),G__54831)){
return f_BANG_.call(null,tx);
} else {
if(cljs.core._EQ_.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [true,true], null),G__54831)){
return f_BANG_.call(null,tx);
} else {
throw (new Error(["No matching clause: ",cljs.core.str.cljs$core$IFn$_invoke$arity$1(G__54831)].join('')));

}
}
}
}
});
hyperfiddle.ide.fiddles.schema_attribute.build_ident_renderer = (function hyperfiddle$ide$fiddles$schema_attribute$build_ident_renderer(special_attrs_state){
return contrib.reagent.from_react_context.call(null,(function (p__54832,value){
var map__54833 = p__54832;
var map__54833__$1 = ((((!((map__54833 == null)))?((((map__54833.cljs$lang$protocol_mask$partition0$ & (64))) || ((cljs.core.PROTOCOL_SENTINEL === map__54833.cljs$core$ISeq$)))?true:false):false))?cljs.core.apply.call(null,cljs.core.hash_map,map__54833):map__54833);
var ctx = cljs.core.get.call(null,map__54833__$1,new cljs.core.Keyword(null,"ctx","ctx",-493610118));
var props = cljs.core.get.call(null,map__54833__$1,new cljs.core.Keyword(null,"props","props",453281727));
return new cljs.core.PersistentVector(null, 5, 5, cljs.core.PersistentVector.EMPTY_NODE, [contrib.reagent.fix_arity_1_with_context,hyperfiddle.ui.hyper_control.call(null,ctx),value,cljs.core.assoc.call(null,ctx,new cljs.core.Keyword(null,"user-with!","user-with!",2055209278),contrib.reactive.partial.call(null,user_with_BANG_,new cljs.core.Keyword(null,"user-with!","user-with!",2055209278).cljs$core$IFn$_invoke$arity$1(ctx),special_attrs_state,ctx)),props], null);
}));
});
hyperfiddle.ide.fiddles.schema_attribute.attrs = new cljs.core.PersistentVector(null, 7, 5, cljs.core.PersistentVector.EMPTY_NODE, [new cljs.core.Keyword("db","ident","db/ident",-737096),new cljs.core.Keyword("db","valueType","db/valueType",1827971944),new cljs.core.Keyword("db","cardinality","db/cardinality",-104975659),new cljs.core.Keyword("db","doc","db/doc",1913350069),new cljs.core.Keyword("db","unique","db/unique",329396388),new cljs.core.Keyword("db","isComponent","db/isComponent",423352398),new cljs.core.Keyword("db","fulltext","db/fulltext",-1432910705)], null);
hyperfiddle.ide.fiddles.schema_attribute.renderer = (function hyperfiddle$ide$fiddles$schema_attribute$renderer(ctx){
var special_attrs_state = contrib.reactive.atom.call(null,null);
var reactive_merge = ((function (special_attrs_state){
return (function (p1__54835_SHARP_){
return hyperfiddle.ide.fiddles.schema_attribute.merge_in_tx.call(null,p1__54835_SHARP_,cljs.core.deref.call(null,special_attrs_state),ctx);
});})(special_attrs_state))
;
var controls = new cljs.core.PersistentArrayMap(null, 3, [new cljs.core.Keyword("db","cardinality","db/cardinality",-104975659),hyperfiddle.ide.fiddles.schema_attribute.build_valueType_and_cardinality_renderer.call(null,special_attrs_state),new cljs.core.Keyword("db","valueType","db/valueType",1827971944),hyperfiddle.ide.fiddles.schema_attribute.build_valueType_and_cardinality_renderer.call(null,special_attrs_state),new cljs.core.Keyword("db","ident","db/ident",-737096),hyperfiddle.ide.fiddles.schema_attribute.build_ident_renderer.call(null,special_attrs_state)], null);
return ((function (special_attrs_state,reactive_merge,controls){
return (function (ctx__$1){
var ctx__$2 = hypercrud.browser.context.with_relations.call(null,cljs.core.update.call(null,cljs.core.dissoc.call(null,ctx__$1,new cljs.core.Keyword(null,"relation","relation",-693437925),new cljs.core.Keyword(null,"relations","relations",-427124442)),new cljs.core.Keyword("hypercrud.browser","result","hypercrud.browser/result",-1521074409),cljs.core.partial.call(null,contrib.reactive.fmap,reactive_merge)));
var result = cljs.core.deref.call(null,new cljs.core.Keyword("hypercrud.browser","result","hypercrud.browser/result",-1521074409).cljs$core$IFn$_invoke$arity$1(ctx__$2));
return cljs.core.into.call(null,new cljs.core.PersistentVector(null, 1, 5, cljs.core.PersistentVector.EMPTY_NODE, [new cljs.core.Keyword(null,"div","div",1057191632)], null),(function (){var iter__9131__auto__ = ((function (ctx__$2,result,special_attrs_state,reactive_merge,controls){
return (function hyperfiddle$ide$fiddles$schema_attribute$renderer_$_iter__54836(s__54837){
return (new cljs.core.LazySeq(null,((function (ctx__$2,result,special_attrs_state,reactive_merge,controls){
return (function (){
var s__54837__$1 = s__54837;
while(true){
var temp__4657__auto__ = cljs.core.seq.call(null,s__54837__$1);
if(temp__4657__auto__){
var s__54837__$2 = temp__4657__auto__;
if(cljs.core.chunked_seq_QMARK_.call(null,s__54837__$2)){
var c__9129__auto__ = cljs.core.chunk_first.call(null,s__54837__$2);
var size__9130__auto__ = cljs.core.count.call(null,c__9129__auto__);
var b__54839 = cljs.core.chunk_buffer.call(null,size__9130__auto__);
if((function (){var i__54838 = (0);
while(true){
if((i__54838 < size__9130__auto__)){
var k = cljs.core._nth.call(null,c__9129__auto__,i__54838);
cljs.core.chunk_append.call(null,b__54839,(function (){var ro = hyperfiddle.ide.fiddles.schema_attribute.read_only_QMARK_.call(null,k,result);
return hyperfiddle.ui.field.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [(0),k], null),ctx__$2,controls.call(null,k),new cljs.core.PersistentArrayMap(null, 1, [new cljs.core.Keyword(null,"read-only","read-only",-191706886),ro], null));
})());

var G__54840 = (i__54838 + (1));
i__54838 = G__54840;
continue;
} else {
return true;
}
break;
}
})()){
return cljs.core.chunk_cons.call(null,cljs.core.chunk.call(null,b__54839),hyperfiddle$ide$fiddles$schema_attribute$renderer_$_iter__54836.call(null,cljs.core.chunk_rest.call(null,s__54837__$2)));
} else {
return cljs.core.chunk_cons.call(null,cljs.core.chunk.call(null,b__54839),null);
}
} else {
var k = cljs.core.first.call(null,s__54837__$2);
return cljs.core.cons.call(null,(function (){var ro = hyperfiddle.ide.fiddles.schema_attribute.read_only_QMARK_.call(null,k,result);
return hyperfiddle.ui.field.call(null,new cljs.core.PersistentVector(null, 2, 5, cljs.core.PersistentVector.EMPTY_NODE, [(0),k], null),ctx__$2,controls.call(null,k),new cljs.core.PersistentArrayMap(null, 1, [new cljs.core.Keyword(null,"read-only","read-only",-191706886),ro], null));
})(),hyperfiddle$ide$fiddles$schema_attribute$renderer_$_iter__54836.call(null,cljs.core.rest.call(null,s__54837__$2)));
}
} else {
return null;
}
break;
}
});})(ctx__$2,result,special_attrs_state,reactive_merge,controls))
,null,null));
});})(ctx__$2,result,special_attrs_state,reactive_merge,controls))
;
return iter__9131__auto__.call(null,hyperfiddle.ide.fiddles.schema_attribute.attrs);
})());
});
;})(special_attrs_state,reactive_merge,controls))
});

//# sourceMappingURL=schema_attribute.js.map
```

**Comment by dustingetz** [comment]
Now, am hitting an existing bug - the options show readonly even though readonly=false is properly set. This is because the @entity is state=loading, which confuses the is-writable check. I guess it eventually reacts and loads the options but the readonly stays true.

**Comment by dustingetz** [comment]
This is the bug. Could it be a load mask issue? I have broken this line many times over and over so am standing down. The attr editor works if you set the ident third and the selects before.
![image](https://user-images.githubusercontent.com/124158/41811908-49ba0118-76e6-11e8-898c-b912a5745fef.png)

**Comment by dustingetz** [comment]
Not fixed in #341 !??
